File: ios_app/MeiliCompanion/Sources/Data/MultiSystem/FollowupModels.swift
import Foundation

/* ===================================================================
 * 回访话术 / 高情商话术(followup-agent)领域模型。
 * android 端对应 `data/model/FollowupModels.kt`。
 *
 * 服务端字段名 1:1 对齐 server.py 真实契约:
 *  - 登录 POST /login/account,密码字段名是 **key** 不是 password;
 *  - JWT 7 天有效、无单设备互踢,401 仅有过期一种(静默重登即可);
 *  - 生成 POST /generate 走 SSE 流式,帧解析在 FollowupRepository。
 * ⚠️ customers 相关 key 服务端本就是 camelCase(savedAt/lastScript),
 *    quota 是 snake(no_monthly_reset)——各自显式 CodingKeys,统一用 plainDecoder。
 * =================================================================== */

/// 任意 JSON 值(顾客档案 data / 战果 business 是自由 dict,服务端原样存取)。
enum JSONValue: Codable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case array([JSONValue])
    case object([String: JSONValue])
    case null

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() { self = .null }
        else if let b = try? c.decode(Bool.self) { self = .bool(b) }
        else if let n = try? c.decode(Double.self) { self = .number(n) }
        else if let s = try? c.decode(String.self) { self = .string(s) }
        else if let a = try? c.decode([JSONValue].self) { self = .array(a) }
        else if let o = try? c.decode([String: JSONValue].self) { self = .object(o) }
        else { self = .null }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .string(let s): try c.encode(s)
        case .number(let n): try c.encode(n)
        case .bool(let b): try c.encode(b)
        case .array(let a): try c.encode(a)
        case .object(let o): try c.encode(o)
        case .null: try c.encodeNil()
        }
    }

    var stringValue: String? { if case .string(let s) = self { return s }; return nil }
    var doubleValue: Double? { if case .number(let n) = self { return n }; return nil }
    var stringList: [String]? {
        if case .array(let a) = self { return a.compactMap(\.stringValue) }
        return nil
    }

    /// 转 JSONSerialization 兼容值(SSE 请求体用)。
    var anyValue: Any {
        switch self {
        case .string(let s): return s
        case .number(let n): return n
        case .bool(let b): return b
        case .array(let a): return a.map(\.anyValue)
        case .object(let o): return o.mapValues(\.anyValue)
        case .null: return NSNull()
        }
    }
}

/// POST /login/account 响应。
struct FuLoginResp: Decodable {
    var token: String?
    var username: String?
    var company: String?
    var error: String?
}

/// GET /quota 响应。groups: 模型组名 → 用量;limit=-1 表示无限。
struct FuQuota: Decodable {
    var username: String?
    var bypass: Bool?
    var noMonthlyReset: Bool?
    var month: String?
    var groups: [String: FuQuotaGroup]?
    var error: String?

    enum CodingKeys: String, CodingKey {
        case username, bypass, month, groups, error
        case noMonthlyReset = "no_monthly_reset"
    }
}

struct FuQuotaGroup: Decodable {
    var used: Int?
    var limit: Int?
}

/// GET /customers 响应(导入历史:按账号隔离的顾客档案+话术版本)。
struct FuCustomers: Decodable {
    var customers: [FuCustomer]?
}

struct FuCustomer: Decodable {
    var name: String?
    var savedAt: String?
    var data: [String: JSONValue]?
    var lastScript: String?
    var favorited: Bool?
    var scripts: [FuScript]?
}

/// 单个历史话术版本。
struct FuScript: Decodable {
    var generatedAt: String?
    var script: String?
    var favorited: Bool?
    var business: [String: JSONValue]?
}

/// GET /employees/list 响应(顾问姓名/门店自动带出)。
struct FuEmployees: Decodable {
    var employees: [FuEmployee]?
}

struct FuEmployee: Decodable {
    var name: String?
    var store: String?
}

File: android_app_v2/app/src/main/java/com/airec/bledemo/data/model/FollowupModels.kt
package com.airec.bledemo.data.model

import com.squareup.moshi.Json

/* ===================================================================
 * 回访话术 / 高情商话术(followup-agent)领域模型。
 *
 * 服务端:huifang-prod /opt/followup-agent(huifang.beautyshining.com)与
 * dev-machine /opt/followup-agent(intelligent.beautyshining.com,高情商变体)。
 * 字段名 1:1 对齐 server.py 真实契约(2026-07-15 上服务器核对):
 *  - 登录 POST /login/account,密码字段名是 **key** 不是 password;
 *  - JWT 7 天有效、无单设备互踢,401 仅有过期一种(静默重登即可);
 *  - 生成 POST /generate 走 SSE 流式(data: {"content": 增量} / data: [DONE]),
 *    SSE 帧解析在 FollowupRepository,不走 Moshi。
 * =================================================================== */

/** POST /login/account 请求体。⚠️ 密码字段名是 key。 */
data class FuLoginReq(
    @Json(name = "username") val username: String,
    @Json(name = "key") val key: String,
)

/** POST /login/account 响应。 */
data class FuLoginResp(
    @Json(name = "token") val token: String? = null,
    @Json(name = "username") val username: String? = null,
    @Json(name = "company") val company: String? = null,
    @Json(name = "error") val error: String? = null,
)

/** GET /quota 响应。groups: 模型组名 → 用量;limit=-1 表示无限。 */
data class FuQuota(
    @Json(name = "username") val username: String? = null,
    @Json(name = "bypass") val bypass: Boolean? = null,
    @Json(name = "no_monthly_reset") val noMonthlyReset: Boolean? = null,
    @Json(name = "month") val month: String? = null,
    @Json(name = "groups") val groups: Map<String, FuQuotaGroup>? = null,
    @Json(name = "error") val error: String? = null,
)

data class FuQuotaGroup(
    @Json(name = "used") val used: Int? = null,
    @Json(name = "limit") val limit: Int? = null,
)

/** GET /customers 响应(导入历史:按账号隔离的顾客档案+话术版本)。 */
data class FuCustomers(
    @Json(name = "customers") val customers: List<FuCustomer>? = null,
)

data class FuCustomer(
    @Json(name = "name") val name: String? = null,
    @Json(name = "savedAt") val savedAt: String? = null,
    @Json(name = "data") val data: Map<String, Any?>? = null,
    @Json(name = "lastScript") val lastScript: String? = null,
    @Json(name = "favorited") val favorited: Boolean? = null,
    @Json(name = "scripts") val scripts: List<FuScript>? = null,
)

/** 单个历史话术版本。 */
data class FuScript(
    @Json(name = "generatedAt") val generatedAt: String? = null,
    @Json(name = "script") val script: String? = null,
    @Json(name = "favorited") val favorited: Boolean? = null,
    @Json(name = "business") val business: Map<String, Any?>? = null,
)

/** GET /employees/list 响应(顾问姓名/门店自动带出)。 */
data class FuEmployees(
    @Json(name = "employees") val employees: List<FuEmployee>? = null,
)

data class FuEmployee(
    @Json(name = "name") val name: String? = null,
    @Json(name = "store") val store: String? = null,
)
